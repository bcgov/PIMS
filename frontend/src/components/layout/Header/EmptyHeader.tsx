import './Header.scss';

import BClogoUrl from 'assets/images/logo-banner.svg';
import PIMSlogo from 'assets/images/PIMSlogo/logo_only.png';
import React from 'react';
import { Nav, Navbar } from 'react-bootstrap';
import styled from 'styled-components';

const VerticalBar = styled.span`
  border-left: 2px solid white;
  font-size: 34px;
  margin: 0 15px 0 25px;
  vertical-align: top;
`;

/**
 * Display an "empty" header bar with limited functionality as a placeholder
 */
const EmptyHeader = () => {
  return (
    <Navbar expand className="App-header">
      <Navbar.Brand className="brand-box">
        <a target="_blank" rel="noopener noreferrer" href="https://www2.gov.bc.ca/gov/content/home">
          <img
            className="bc-gov-icon"
            src={BClogoUrl}
            width="156"
            height="43"
            alt="Government of BC logo"
          />
        </a>
        <VerticalBar />
        <img className="pims-logo" src={PIMSlogo} height="50" alt="PIMS logo" />
      </Navbar.Brand>
      <Nav className="title mr-auto">
        <Nav.Item>
          <h1 className="longAppName">Property Inventory Management System</h1>
          <h1 className="shortAppName">PIMS</h1>
        </Nav.Item>
      </Nav>
    </Navbar>
  );
};

export default EmptyHeader;
