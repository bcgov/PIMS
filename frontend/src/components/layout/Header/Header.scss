@import '../../../fonts.scss';

.App-header.navbar {
  padding: 0 10px;
  min-height: 70px;
  color: #ffffff;

  .longAppName {
    display: none;
  }
  .shortAppName {
    display: block;
  }

  .brand-box {
    padding: 10px 0;

    .pims-logo {
      margin: 0 10px;
    }
  }

  .title h1 {
    margin-top: 10px;
    padding-left: 0px;
    text-align: center;
    font-size: 24px;
    text-decoration: none solid rgb(255, 255, 255);
    font-family: 'BCSans', Fallback, sans-serif;
    font-weight: 700;
    white-space: normal;
  }

  .other {
    display: flex;
    flex-direction: row-reverse;
    align-items: center;
    padding: 0;
    :hover {
      color: red;
      cursor: pointer;
    }
  }
}
.modal-content .label {
  font-weight: 700;
}

// show long App Name when space allows
@media (min-width: 992px) {
  .App-header.navbar {
    .longAppName {
      display: block;
      text-align: left;
      padding-left: 40px;
    }
    .shortAppName {
      display: none;
    }
  }
}
