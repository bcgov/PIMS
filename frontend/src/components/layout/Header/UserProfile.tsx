import variables from '_variables.module.scss';
import { ILookupCode } from 'actions/ILookupCode';
import profileUrl from 'assets/images/profile.svg';
import * as API from 'constants/API';
import { useConfiguration } from 'hooks/useConfiguration';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import useCodeLookups from 'hooks/useLookupCodes';
import React from 'react';
import { Image, NavDropdown } from 'react-bootstrap';
import { FaSignOutAlt } from 'react-icons/fa';
import styled from 'styled-components';

/** the styling for the dropdown menu that appears after clicking the user's name */
const StyleDropDown = styled(NavDropdown)`
  font-size: 14px;
  .dropdown-menu {
    width: 300px;
    height: 150px;
  }
  .nav-link {
    color: #fff;
    padding: 1px;
  }
  .dropdown-item {
    background-color: ${variables.primaryLightColor};
    border-top: 2px solid ${variables.accentColor};
  }
`;

/** shaded box the users system roles will be displayed in */
const RolesBox = styled.div`
  background-color: ${variables.filterBackgroundColor};
  margin: 5px;
`;

/** the text contained in the logout footer  */
const LogoutText = styled.p`
  color: #fff;
  margin-top: 0.5rem;
  margin-left: 120px;
`;

/** the styling for the avatar next to user's name */
const ProfileAvatar = styled(Image)`
  height: 30px;
  width: 30px;
`;

/** the styling for the logout icon in the logout footer */
const LogoutButton = styled(FaSignOutAlt)`
  margin-bottom: 2px;
  margin-left: 5px;
`;

/** Component that allows the user to logout, and gives information on current user's agency/roles */
export const UserProfile: React.FC = () => {
  const keycloak = useKeycloakWrapper();
  const displayName =
    keycloak.displayName ??
    (!!keycloak.firstName && !!keycloak.lastName
      ? `${keycloak.firstName} ${keycloak.lastName}`
      : 'default');
  const configuration = useConfiguration();
  const lookupCodes = useCodeLookups();
  const agencyOptions = lookupCodes.getByType(API.AGENCY_CODE_SET_NAME);
  const roles = keycloak.roles.join(', ');

  return (
    <>
      <ProfileAvatar src={profileUrl} rounded />
      <StyleDropDown className="px-0" title={displayName} id="user-dropdown">
        <p style={{ margin: 5 }}>
          <b>
            {
              agencyOptions.find(
                (x: ILookupCode) => x.id.toString() === keycloak.agencyId?.toString(),
              )?.name
            }
          </b>
        </p>
        <RolesBox>
          <p style={{ margin: 5 }}>
            <b>
              System Role(s):
              <br />
            </b>
            {roles}
          </p>
        </RolesBox>
        <NavDropdown.Item
          onClick={() => {
            keycloak.obj!.logout({ redirectUri: `${configuration.baseUrl}/logout` });
          }}
        >
          <LogoutText>
            Log out of PIMS <LogoutButton />
          </LogoutText>
        </NavDropdown.Item>
      </StyleDropDown>
    </>
  );
};
