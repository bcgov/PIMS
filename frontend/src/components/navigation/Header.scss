.bar {
  padding: 5px;
  background-color: #036;
  border-bottom: 2px solid #fcba19;
  color: #fff;
  height: 70px;
  top: 0px;
  width: 100%;
  white-space: nowrap;

  h1 {
    font-family: Arial;
    font-size: 24px;
    color: #ffffff;
    text-decoration: none solid rgb(255, 255, 255);
    font-family: "BCSans" Fallback, sans-serif;
    font-weight: 700;
    overflow: hidden;
  }

  a {
    width: 30%;
    text-align: right;
    padding-right: 27px; 
  }

  .title {
    margin-top: 15px;
    width: 70%;
    text-align: left;
  }

  h2 {
    margin-top: 15px;
    padding-right: 20px;
    font-family: Arial;
    font-size: 20px;
    color: #ffffff;
    text-decoration: none solid rgb(255, 255, 255);
    overflow: hidden;
  }

  .banner {
    display: inline-flex;
    width: 100%;
    /* border-style: dotted;
    border-width: 1px;
    border-color: lightgrey; */
  }

  .exit {
    height: 68px;
    width: 70px;
    flex-flow: column;
    align-items: center;
    background-color: #38598a;
    :hover {
      cursor: pointer;
    }
  }

  p {
    width: 49px;
    height: 18px;
    font-family: Arial;
    font-size: 16px;
    color: #ffffff;
    text-decoration: none solid rgb(255, 255, 255);
  }

  .other {
    flex-grow: 1;
    justify-content: flex-end;
    align-items: center;
    /* border-style: dotted;
    border-width: 1px;
    border-color: lightgrey; */
  }

  p {
    padding-right: 50px;
    margin: 0;
  }

  #user-list-dropdown {
    width: 240px;
    height: 64px;
    margin-right: 16px;
    background-color: #fcba19;
    font-family: Arial;
    font-size: 20px;
    color: #ffffff;
    text-decoration: none solid rgb(255, 255, 255);
    text-align: left;
    overflow: visible;
  }
}
