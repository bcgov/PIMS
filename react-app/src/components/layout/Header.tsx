import React from 'react';
import headerImageLarge from '@/assets/images/BCID_H_rgb_pos.png';
import headerImageSmall from '@/assets/images/BCID_V_rgb_pos.png';
import { AppBar, Link, Box, Button, Divider, Toolbar, Typography, useTheme } from '@mui/material';
import { Link as RouterLink } from 'react-router-dom';
import { useKeycloak } from '@bcgov/citz-imb-kc-react';

const AppBrand = () => {
  const theme = useTheme();
  return (
    <Box
      sx={{
        '& a': {
          display: 'flex',
          alignItems: 'center',
          overflow: 'hidden',
          textDecoration: 'none',
        },
        '& img': {
          mr: '12px',
        },
      }}
    >
      <RouterLink to="/" aria-label="Go to PIMS Home">
        <picture>
          <source srcSet={headerImageLarge} media="(min-width: 960px)"></source>
          <source srcSet={headerImageSmall} media="(min-width: 600px)"></source>
          <img
            src={headerImageSmall}
            alt={'Government of British Columbia'}
            style={{ height: '48px' }}
          />
        </picture>
        <Divider
          style={{ borderBottomWidth: '26px', borderRightWidth: '2px' }}
          orientation="vertical"
        />
        <Typography marginLeft={'16px'} color={theme.palette.black.main} variant="h5">
          Property Inventory Management System
        </Typography>
      </RouterLink>
    </Box>
  );
};

const Header: React.FC = () => {
  const { logout, isAuthenticated, login } = useKeycloak();
  const theme = useTheme();
  const handleLoginButton = () => {
    if (isAuthenticated) {
      logout();
    } else {
      login({ idpHint: 'idir' });
    }
  };
  return (
    <AppBar
      elevation={0}
      style={{
        backgroundColor: theme.palette.white.main,
        height: '74px',
        display: 'flex',
        position: 'relative',
        justifyContent: 'center',
        borderBottom: '1px solid',
        borderBottomColor: theme.palette.gray.main,
      }}
    >
      <Toolbar>
        <AppBrand />
        <Box flexGrow={1}></Box>
        <Box textAlign={'center'} alignItems={'center'} gap={'32px'} display={'flex'}>
          {isAuthenticated && (
            <>
              <Link underline="none" href="#" variant="h5">
                Active Inventory
              </Link>
              <Link underline="none" href="#" variant="h5">
                Disposal Inventory
              </Link>
              <Link underline="none" href="/admin/users" variant="h5">
                Users
              </Link>
            </>
          )}
          <Button onClick={() => handleLoginButton()} color="secondary" variant="contained">
            {isAuthenticated ? 'Logout' : 'Login'}
          </Button>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default Header;
