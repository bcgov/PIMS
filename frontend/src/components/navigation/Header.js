import { useKeycloak } from 'react-keycloak';
import React, { useState, useEffect } from 'react';
import { Image, DropdownButton, Dropdown } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import './Header.scss';
import logoUrl from './logo-banner.svg';
import profileUrl from './profile.svg';
import logout from './logout.svg';
import dropdown from './dropdown.svg';
import API from '../../utils/API';

function Header(props) {
  const history = useHistory();
  const [isAdmin, setAdmin] = useState(false);
  const [activeUser, setActiveUser] = useState({ id: "" });
  const [users, setUsers] = useState([]);
  const { keycloak } = useKeycloak();
  useEffect(() => {
    async function fetchData() {
      if (keycloak.authenticated) {
        API.token = keycloak.token
        if (keycloak && keycloak.realmAccess && keycloak.realmAccess.roles) {
          const isAdmin = keycloak.realmAccess.roles.indexOf("administrator") >= 0;
          setAdmin(isAdmin);
          if (isAdmin) {
            const allUsers = await API.getAllUsers();
            if (allUsers) {
              allUsers.push({ firstName: "All", lastName: "Users", id: "all" });
            }
            setUsers(allUsers);
            props.setActiveUserId(keycloak.subject);
          }
        }

        keycloak.loadUserInfo().then(() => {
          //for simplicity just map the keycloak user details
          setActiveUser({ id: keycloak.subject, firstName: keycloak.userInfo.given_name, lastName: keycloak.userInfo.family_name});
        });
      }
    }
    fetchData();
  }, [keycloak]);

  const selectUser = (eventKey, event) => {
    const matchingUser = users.find(user => user.id === eventKey);
    if (!matchingUser) {
      throw Error("clicked on user not in user list.");
    }
    setActiveUser(matchingUser);
    props.setActiveUserId(matchingUser.id);
  }

  return (
    <div className="Logout">
      <div className="banner">
        <a href="https://gov.bc.ca" alt="British Columbia">
          <img
            src={logoUrl}
            width="156"
            height="43"
            alt="Go to the Government of British Columbia website"
          />
        </a>
        <h1>Geo-spatial Real Estate Inventory</h1>
      </div>
      {!!keycloak.authenticated ? (<div className="other">
        {!!isAdmin ? (<DropdownButton title={<span><img
          src={dropdown}
          width="49"
          height="49"
          alt="map pin"
        />Filter</span>} variant="filter" id="user-list-dropdown">
          {users.map((item, idx) => (
            <Dropdown.Item eventKey={item.id} key={idx} active={item.id === activeUser.id} onSelect={selectUser}>{item.firstName} {item.lastName}</Dropdown.Item>
          ))}
        </DropdownButton>) : <span></span>}

        <h2><img
          src={profileUrl}
          width="42"
          height="42"
          alt="profile"
        />{activeUser.firstName} {activeUser.lastName}
        </h2>
        <div className="exit" onClick={() => {
          history.push('/');
          keycloak.logout();
        }}>
          <Image src={logout} width="34" height="34" alt="exit"></Image>
          <p>Sign-out</p>
        </div>

      </div>) : (<span></span>)}
    </div>
  );
}

export default Header;


