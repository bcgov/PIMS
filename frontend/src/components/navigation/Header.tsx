import { useKeycloak } from 'react-keycloak';
import React from 'react';
import { Image, } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';
import './Header.scss';
import logoUrl from './logo-banner.svg';
import logout from './logout.svg';

function Header() {
  const history = useHistory();
  const { keycloak } = useKeycloak();
  //example of how to check role: const isAdmin = keycloak?.realmAccess?.roles.includes(USER_ROLES[Permission.ADMIN]) == true;
  return (
    <header className="App-header">
      <div className="bar">
        <div className="banner">
          <a href="https://gov.bc.ca" className="bc-gov-icon">
            <img
              src={logoUrl}
              width="156"
              height="43"
              alt="Go to the Government of British Columbia website"
            />
          </a>
          <h1 className="title">Property Inventory Management System</h1>
        </div>
        {!!keycloak.authenticated ? (<div className="other">
          <div className="exit" onClick={() => {
            history.push('/');
            keycloak.logout();
          }}>
            <Image src={logout} width="34" height="34" alt="exit"></Image>
            <p>Sign-out</p>
          </div>

        </div>) : (<span></span>)}
      </div>
    </header>
  );
}

export default Header;


