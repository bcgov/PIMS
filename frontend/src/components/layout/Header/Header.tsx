import './Header.scss';

import React, { useState } from 'react';
import { Navbar, Row, Col, Modal, Button, Nav } from 'react-bootstrap';
import BClogoUrl from 'assets/images/logo-banner.svg';
import PIMSlogo from 'assets/images/PIMSlogo/logo_only.png';
import { useHistory } from 'react-router-dom';
import { IGenericNetworkAction, clear } from 'actions/genericActions';
import { RootState } from 'reducers/rootReducer';
import { useSelector, useDispatch } from 'react-redux';
import { FaBomb } from 'react-icons/fa';
import _ from 'lodash';
import { UserProfile } from './UserProfile';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import styled from 'styled-components';

const VerticalBar = styled.span`
  border-left: 2px solid white;
  font-size: 34px;
  margin: 0 15px 0 25px;
  vertical-align: top;
`;

const Header = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const keycloak = useKeycloakWrapper();
  if (history.location.pathname === '/') {
    history.replace('/mapview');
  }
  const [show, setShow] = useState(false);
  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);
  const handleClear = () => {
    errors.forEach(error => dispatch(clear(error.name)));
    setShow(false);
  };

  const isNetworkError = (x: any): x is IGenericNetworkAction =>
    (x as IGenericNetworkAction).type === 'ERROR';
  const errors = useSelector<RootState, IGenericNetworkAction[]>(state => {
    const errors: IGenericNetworkAction[] = [];
    _.values(state).forEach(reducer => {
      _.values(reducer)
        .filter(x => x instanceof Object)
        .forEach(action => {
          if (isNetworkError(action)) {
            errors.push(action);
          }
        });
    });
    return errors;
  });
  //TODO: styling - this is a placeholder, need UI.
  const errorModal = (errors: IGenericNetworkAction[]) => (
    <Modal show={show} onHide={handleClose}>
      <Modal.Header closeButton>
        <Modal.Title>Errors</Modal.Title>
      </Modal.Header>

      <Modal.Body style={{ maxHeight: '500px', overflowY: 'scroll' }}>
        {errors.map((error: IGenericNetworkAction, index: number) => (
          <Row key={index} style={{ wordBreak: 'break-all' }}>
            {process.env.NODE_ENV === 'development' ? (
              <Col>
                <abbr title={error.error?.response?.config?.url}>
                  {error.error?.response?.config?.url?.substr(0, 20)}
                </abbr>
                : {error.error?.response?.statusText} data:{' '}
                {JSON.stringify(error.error?.response?.data)}
              </Col>
            ) : (
              <Col>
                <abbr title={error.error?.response?.config?.url}>
                  {error.error?.response?.config?.url?.substr(0, 20)}
                </abbr>
                : ({error.error?.response?.statusText ?? 'unknown'}){' '}
                {error.error?.response?.data?.error}
              </Col>
            )}
          </Row>
        ))}
      </Modal.Body>

      <Modal.Footer>
        <Button variant="primary" onClick={handleClear}>
          Close & Clear Errors
        </Button>
      </Modal.Footer>
    </Modal>
  );
  return (
    <Navbar expand className="App-header">
      <Navbar.Brand className="brand-box">
        <a target="_blank" rel="noopener noreferrer" href="https://www2.gov.bc.ca/gov/content/home">
          <img
            className="bc-gov-icon"
            src={BClogoUrl}
            width="156"
            height="43"
            alt="Go to the Government of British Columbia website"
          />
        </a>
        <VerticalBar />
        <img className="pims-logo" src={PIMSlogo} height="50" alt="PIMS logo" />
      </Navbar.Brand>
      <Nav className="title mr-auto">
        <Nav.Item>
          <h1 className="longAppName">Property Inventory Management System</h1>
          <h1 className="shortAppName">PIMS</h1>
        </Nav.Item>
      </Nav>
      {keycloak.obj.authenticated && <UserProfile />}
      <Nav className="other">
        {errors && errors.length ? (
          <FaBomb size={30} className="errors" onClick={handleShow} />
        ) : null}
      </Nav>
      {errorModal(errors)}
    </Navbar>
  );
};

export default Header;
