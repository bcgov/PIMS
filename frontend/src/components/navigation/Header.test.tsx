import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { useKeycloak } from '@react-keycloak/web';
import Header from './Header';

jest.mock('@react-keycloak/web');

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

test('header renders correctly', () => {
  (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { authenticated: false } });
  const tree = renderer.create(<Header />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('authenticated users are shown the logout button', () => {
  (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { authenticated: true } });
  const { getByText } = render(<Header />);
  expect(getByText('Sign-out')).not.toBeNull();
});

test('unAuthenticated users are not shown the logout button', () => {
  (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { authenticated: false } });
  const { queryByText } = render(<Header />);
  expect(queryByText(/sign-out/i)).toBeNull();
});

test('the logout button calls keycloaks logout() method', () => {
  const logout = jest.fn();
  (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { logout: logout, authenticated: true } });

  const { getByText } = render(<Header />);
  fireEvent.click(getByText('Sign-out'));

  expect(logout.mock.calls.length).toBe(1);
});
