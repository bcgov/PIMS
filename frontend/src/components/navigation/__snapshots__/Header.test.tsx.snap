// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header renders correctly 1`] = `
<nav
  className="App-header navbar navbar-expand-xl navbar-light fixed-top"
>
  <div
    className="bar container-fluid"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="brand-box row"
        >
          <div
            className="col-lg-2 col-md-2"
          />
          <div
            className="brand col-lg-1 col-md-4 col-sm-6 col-8"
          >
            <span
              className="navbar-brand"
              onClick={[Function]}
            >
              <img
                alt="Go to the Government of British Columbia website"
                className="bc-gov-icon"
                height="43"
                src="logo-banner.svg"
                width="156"
              />
            </span>
          </div>
          <div
            className="title col-lg-7 col-md-2 col-sm-4 col-3"
          >
            <h1
              className="longAppName"
            >
              Property Inventory Management System
            </h1>
            <h1
              className="shortAppName"
            >
              PIMS
            </h1>
          </div>
        </div>
      </div>
      <div
        className="other col-lg-1 col-md-3 col-sm-3 col-3"
      />
    </div>
  </div>
</nav>
`;
